analyzers/src/SonarAnalyzer.CSharp/Rules/InfiniteRecursion.RoslynCfg.cs

```diff
@@ -0,0 +1,179 @@
+                    SyntaxKindEx.LocalFunctionStatement, SyntaxKind.SimpleLambdaExpression, SyntaxKind.AnonymousMethodExpression, SyntaxKind.ParenthesizedLambdaExpression
+                };
+                if (c.Node.IsKind(SyntaxKindEx.LocalFunctionStatement))
+                {
+                    // we need to go up and track all possible enclosing local function statements
+                    foreach (var enclosingFunction in c.Node.Ancestors().Where(x => enclosingKinds.Contains(x.Kind())).Reverse())
+                    {
+                        if (enclosingFunction.IsKind(SyntaxKindEx.LocalFunctionStatement))
+                        {
+                            cfg = cfg.GetLocalFunctionControlFlowGraph(c.SemanticModel.GetDeclaredSymbol(enclosingFunction) as IMethodSymbol);
+                        }
+                        else
+                        {
+                            var operationWrapper = cfg.FlowAnonymousFunctionOperations().Single(x => x.WrappedOperation.Syntax == enclosingFunction);
+                            cfg = cfg.GetAnonymousFunctionControlFlowGraph(operationWrapper);
+                        }
+                    }
+
+                    cfg = cfg.GetLocalFunctionControlFlowGraph(symbol as IMethodSymbol);
+                }
+
+                var walker = new CommonRecursionSearcher(new RecursionContext<ControlFlowGraph>(cfg, symbol, identifier.GetLocation(), c, "method's recursion"));
+                if (!walker.CheckPaths() && !cfg.ExitBlock.IsReachable)
+                {
+                    c.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, identifier.GetLocation(), "method's recursion"));
+                }
+            }
+
+            private class CommonRecursionSearcher : CfgAllPathValidator
+            {
+                private readonly ISymbol memberToCheck;
+                private readonly Action reportIssue;
+                private readonly ControlFlowGraph cfg;
+                private readonly bool isGetAccesor;
+
+                public CommonRecursionSearcher(RecursionContext<ControlFlowGraph> context, bool isGet = true)
+                    : base(context.ControlFlowGraph)
+                {
+                    memberToCheck = context.AnalyzedSymbol;
+                    reportIssue = context.ReportIssue;
+                    cfg = context.ControlFlowGraph;
+                    isGetAccesor = isGet;
+                }
+
+                public bool CheckPaths()
+                {
+                    if (CheckAllPaths())
+                    {
+                        reportIssue();
+                        return true;
+                    }
+
+                    return false;
+                }
+
+                protected override bool IsValid(BasicBlock block) =>
+                    FindReferenceToSelf(block);
+
+                protected override bool IsInvalid(BasicBlock block) => false;
+
+                private bool FindReferenceToSelf(BasicBlock block)
```

Comment:
While looking at the xor logic and `bool` result, this name doesn't help too much with the understanding. Can it be improved somehow?

Mechanically, let's start the discussion with this proposal. But I feel that we should try to explain what kind of reference. Probably not read, not write. Maybe something like useful reference (don't like it, I can't find the word)
```suggestion
                private bool ContainsReferenceToSelf(BasicBlock block)
```

Reply:
We no longer need it, since we've embedded the logic to IsValid method